packages/product-json-to-csv/src/map-product-data.js

```diff
@@ -74,92 +102,258 @@ export default class ProductMapping {
+    const languages: Array<string> = labels
+      .map(Object.keys) // returns array of arrays with language keys
+      // flatten subarrays to one big array
+      .reduce((all, langKeys) => {
+        all.push(...langKeys)
+        return all
+      }, [])
+      // filter out duplicates
+      .filter((value, index, self) => self.indexOf(value) === index)
+
+    languages.forEach(lang => {
+      const headerKey: string = `${name}.${lang}`
+      const values = []
+
+      labels.forEach(label => {
+        values.push(label[lang] || '')
+      })
+
+      mappedValues[headerKey] = values.join(this.multiValDel)
+    })
+
+    return mappedValues
+  }
+
+  _mapSetAttribute(name: string, values: Array<any>) {
+    const [firstValue] = values
+    let mappedValue: string = ''
+
+    // empty set
+    if (values.length === 0) mappedValue = ''
+
+    // string, boolean, number
+    if (!isObject(firstValue)) mappedValue = values.join(this.multiValDel)
+    else if (firstValue.id && firstValue.typeId)
+      // reference
+      mappedValue = values
+        .map(value => ProductMapping._mapReference(value))
+        .join(this.multiValDel)
+    else if (firstValue.key && !isObject(firstValue.label))
+      // enum
+      mappedValue = map(values, 'key').join(this.multiValDel)
+    else if (firstValue.key && isObject(firstValue.label)) {
+      // lenum
+      const labels = map(values, 'label')
+      // map all language labels
+      const mappedValues = this._mapLabelsToAllAttributes(name, labels)
+      // add lenum keys as a main attribute value
+      mappedValues[name] = map(values, 'key').join(this.multiValDel)
+      return mappedValues
+    } else {
+      // ltext
+      const mappedValues = this._mapLabelsToAllAttributes(name, values)
+      // copy value from selected language as a main attribute value
+      mappedValues[name] = mappedValues[`${name}.${this.lang}`]
+      return mappedValues
+    }
+
+    return {
+      [name]: mappedValue,
+    }
+  }
+
+  /**
+   * Method will take an attribute and map it to string
+   * @param attribute Attribute which should be mapped
+   * @param allAttributes Result object with all attributes
+   * @returns {*} Mapped attribute
+   * @private
+   */
+  _mapAttribute(attribute: Attribute): Object {
+    const { name, value } = attribute
+    let mappedAttribute: Object = {}
+
+    if (isObject(value) && !Array.isArray(value)) {
+      // ltext, enum, lenum
+      if (value.id && value.typeId)
+        // reference
+        mappedAttribute[name] = ProductMapping._mapReference(value)
+      else if (value.key && !isUndefined(value.label))
+        // ENUM or LENUM attribute
+        mappedAttribute = ProductMapping._mapLenumOrEnumAttribute(name, value)
+      else
+        // LTEXT attribute
+        mappedAttribute = this._mapLtextAttribute(name, value)
+    } else if (Array.isArray(value)) {
+      // SET attribute
+      mappedAttribute = this._mapSetAttribute(name, value)
```

Comment:
The reason why the simple `.join(delim)` does not work there is, that the `Product.mapAttributeValue` can return an object instead of a string (eg for `ltext` it returns an object with all translations)

But I will refactor it .. I agree with you, that we should not duplicate the code

Reply:
I refactored mapping of set values here 45d7650 .. It looks much better now..